# Introduction
in this part, we want to know similarity between 2 words, there some problem in this case, 
1. make sure both of words is same phonetic or not by using metaphone algorithm
2. do similiraty with levensthein similarity

before we start this simple script, let's intoduce what library we need,
1. Pandas to show dataframe
2. Jellysih to show phonetic algorithm, metaphone and soundex
3. textdistance to show similiraity between 2 word

In [89]:
#install library what we needed
!pip install jellyfish
!pip install textdistance

In [90]:
#import library what we needed
import textdistance
import jellyfish

let's make def to make simple code for this case
siphonetic, it means we using phonetic score and levenshtein score
levens, it means we only use levenshtein score

In [91]:

def siphonetic(row):
    kata1 = str(row['kata1'])
    kata2 = str(row['kata2'])
    if jellyfish.metaphone(kata1)==jellyfish.metaphone(kata2):
        scoremetaphone=1
    else:
        scoremetaphone=0
        
    if jellyfish.soundex(kata1)==jellyfish.soundex(kata2):
        scoresoundex=1
    else:
        scoresoundex=0

    phonetic=(scoresoundex+scoremetaphone)/2
    score_levenshtein=textdistance.levenshtein.normalized_similarity(kata1, kata2)
    
    score_similarity=(score_levenshtein+phonetic)/2
    return score_similarity

def levens(row):
    kata1 = str(row['kata1'])
    kata2 = str(row['kata2'])
    return textdistance.levenshtein.normalized_similarity(kata1, kata2)
    

### Lets try to excute this logic 
we excute this simple logic to typo dataset from google searching. <br>
there are 39172 typo words and also the right words

In [97]:
import pandas as pd
typo=pd.read_csv('C://Users//Lenovo//Downloads//typo-corpus-r1.txt',sep='\t',header=None)
typo=typo[[0,1]]
typo.columns=['kata1','kata2']

In [95]:
%%time
typo['siphonetic']=typo.apply(lambda x: siphonetic(x),axis=1)

Wall time: 6.15 s


In [96]:
%%time
typo['levens']=typo.apply(lambda x: levens(x),axis=1)

Wall time: 4.31 s


based on excute time, siphonetic need time less than levens. <br>
it's okay, because on siphonetic we using 2 phonetic algorithm, and levens as similarity algoritm 

In [92]:
typo.head(50)

kata1   kata2  siphonetic    levens
0      york    work    0.375000  0.750000
1    prople  people    0.416667  0.833333
2     hring  hiring    0.916667  0.833333
3    prople  people    0.416667  0.833333
4    prople  people    0.416667  0.833333
5     nrews    news    0.400000  0.800000
6     newss    news    0.900000  0.800000
7       onn      on    0.833333  0.666667
8       nes    news    0.875000  0.750000
9      ther     the    0.375000  0.750000
10      onn      on    0.833333  0.666667
11      onn      on    0.833333  0.666667
12       oj      on    0.250000  0.500000
13    wavve    wave    0.900000  0.800000
14      wve    wave    0.625000  0.750000
15       ak       a    0.250000  0.500000
16       sa       a    0.250000  0.500000
17    wavie    wave    0.900000  0.800000
18     waze    wave    0.375000  0.750000
19       ia       a    0.250000  0.500000
20   googel  google    0.583333  0.666667
21       aj       a    0.250000  0.500000
22     wavw    wave    0.875000  0.750000
23       za       a    0.250000  0.500000
24    awave    wave    0.400000  0.800000
25  googlee  google    0.928571  0.857143
26     wabe    wave    0.625000  0.750000
27     wvae    wave    0.500000  0.500000
28   goolge  google    0.333333  0.666667
29     wavd    wave    0.375000  0.750000
30       ya       a    0.250000  0.500000
31     wava    wave    0.875000  0.750000
32       aa       a    0.500000  0.500000
33       wa       a    0.250000  0.500000
34    googe  google    0.416667  0.833333
35    goole  google    0.416667  0.833333
36       ja       a    0.250000  0.500000
37       af       a    0.250000  0.500000
38     youh     you    0.875000  0.750000
39      frm    from    0.875000  0.750000
40   invote  invite    0.916667  0.833333
41  iinvite  invite    0.678571  0.857143
42   googel  google    0.583333  0.666667
43    invie  invite    0.416667  0.833333
44    awave    wave    0.400000  0.800000
45     waze    wave    0.375000  0.750000
46    wavve    wave    0.900000  0.800000
47   invitr  invite    0.666667  0.833333
48      wve    wave    0.625000  0.750000
49      ave    wave    0.375000  0.750000

## let's take simple case
if <br>
word1 = **york** <br>
word2 = **work** <br>
siphonetic score given **0.375**, <br>
because the score saw phonetic of "york" and "work" is different, even levens score is 0.75, but if phonetic score is low, its make score of siphonetic is low <br>

While levens given score 0.75 <br>


lets other case <br>
word1= **"hring"** <br>
word2= **hiring** <br>

score siphonetic is 0.916667 <br>	score levens is 0.833333 <br>
its happen because, between **hring** and **hiring** have same phoentic score, so it makes the score siphonetic more than score levens

